components/admin_console/billing/company_info_edit.tsx

```diff
@@ -1,17 +1,186 @@
 
 type Props = {
 
 };
 
 const CompanyInfoEdit: React.FC<Props> = () => {
+    const dispatch = useDispatch();
+    let companyInfo = useSelector((state: GlobalState) => state.entities.cloud.customer);
+
+    if (!companyInfo) {
+        companyInfo = {
+            name: 'Test Company',
+            num_employees: 100,
+            company_address: {
+                line1: '123 Main St',
+                city: 'Anytown',
+                state: 'ON',
+                country: 'CA',
+                postal_code: 'H0H 0H0',
+            },
+        };
+    }
+
+    const [companyName, setCompanyName] = useState(companyInfo?.name);
+    const [numEmployees, setNumEmployees] = useState(companyInfo?.num_employees);
+
+    const [address, setAddress] = useState(companyInfo?.company_address?.line1);
+    const [address2, setAddress2] = useState(companyInfo?.company_address?.line2);
+    const [city, setCity] = useState(companyInfo?.company_address?.city);
+    const [postalCode, setPostalCode] = useState(companyInfo?.company_address?.postal_code);
+    const [country, setCountry] = useState(getName(companyInfo?.company_address?.country || 'US'));
+    const [state, setState] = useState(companyInfo?.company_address?.state);
+
+    const [sameAsBillingAddress, setSameAsBillingAddress] = useState(!companyInfo?.company_address?.line1);
+    const [isValid, setIsValid] = useState<boolean | undefined>(undefined);
+    const [isSaving, setIsSaving] = useState(false);
+
+    const setValidation = () => {
+        if (sameAsBillingAddress) {
+            setIsValid(Boolean(companyName));
+        } else {
+            setIsValid(Boolean(companyName && address && city && postalCode && country && state));
+        }
+    };
+
+    const updateState = (setStateFunc: (value: any) => void) => {
+        return (event: React.ChangeEvent<HTMLInputElement>) => {
+            setStateFunc(event.target.value);
+            setValidation();
+        };
+    };
+
+    useEffect(() => {
+        dispatch(getCloudCustomer());
+    }, []);
+
+    useEffect(() => {
+        setValidation();
+    }, [sameAsBillingAddress, companyName, address, city, postalCode, country, state]);
+
+    const handleSubmit = async () => {
+        setIsSaving(true);
+        await dispatch(updateCloudCustomer({name: companyName, num_employees: parseInt(numEmployees, 10)}));
+
+        if (sameAsBillingAddress) {
+            await dispatch(updateCloudCustomerAddress({
+                city: '',
+                country: '',
+                line1: '',
+                line2: '',
+                postal_code: '',
+                state: '',
+            }));
+        } else {
+            await dispatch(updateCloudCustomerAddress({
+                city: city || '',
+                country: country || '',
+                line1: address || '',
+                line2: address2 || '',
+                postal_code: postalCode || '',
+                state: state || '',
```

Comment:
Are we checking that these fields are required? I believe they all should be expect `line2`

Reply:
They should all be required and we're checking as part of the validation step, but because their definition in `useState` has them potentially null (since both `billing_address` and `company_address` can potentially be null), each of those types are `string | undefined` and need to be null checked. So this is more done for the sake of TypeScript.